ramips: wevo: set uImage name and status led

mans0n opened the pull request:
~~- set uImage name similar to that of stock FW, so that the stock bootloader will not complain.~~
~~- remove WAN port index.~~
~~- load WAN MAC address directly from the flash.~~
~~- provide label MAC address.~~
~~- use USB LED for boot/failsafe/upgrade status indication.~~
~~- replace gpio-keys-polled with gpio-keys.~~
~~- increase flash's SPI frequency to 80MHz.~~
~~- add mt76 led nodes to make WiFi LEDs work.~~
~~- drop unnecessary pinmux groups.~~

The stock firmware and bootloader only accept uImage with names that match
certain patterns. This patch enables OpenWrt installation from stock
firmware without having to reflash the bootloader or access the UART
console.

Installation via web interface:
1.  Flash **initramfs** image through the stock web interface.
2.  Boot into OpenWrt and perform sysupgrade with sysupgrade image.

And also make use of the USB LED for boot/failsafe/upgrade status
indication. After boot it returns to its original role.

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -395,7 +393,9 @@ ramips_setup_interfaces()
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
 		ucidef_add_switch "switch0" \
 			"0:lan:4" "1:lan:3" "2:lan:2" "3:lan:1" "4:wan" "6@eth0"
```
Won't this result in a collision for port 4 in luci?

mans0n replied:
Hmm it seemed okay when I tested...

adrianschmutzler replied:
Okay, I've also found other similar entries in 02_network.
However, let me just change my question: What is your reason to remove the ":5" for the wan port?

mans0n replied:
AFAIK port indices are only used for sorting ethernet ports in the same role.
https://github.com/openwrt/luci/blob/master/modules/luci-base/htdocs/luci-static/resources/network.js#L541
For this model, there is only one WAN port, and the index number "5" does not affect in any way, for now.
Still it can be leaved if you think we can use port index in some other ways in the future.

adrianschmutzler replied:
I see your point. Still, if the ":5" has no negative effect, removing it looks to me like removing information with no reason for it. So I'd keep it.
However, to be honest I'm aware that the subject isn't worth the big discussion we are having here. So I will also be able to sleep at night if you remove it :-)

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/02_network:
```diff
@@ -510,7 +510,9 @@ ramips_setup_macs()
+	netgear,wnce2001|\
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
 		wan_mac=$(mtd_get_mac_binary factory 0x2e)
```
The looks strange to me, as &ethernet pulls the address from <&factory 0xe000>.

Can you provide a hexdump of the whole factory partition (for both devices if possible)?

mans0n replied:
I'm not sure I can post the whole partition here.
The MAC addresses of 11acnas device I have are xx:xx:xx:99:xx:yy,
and I searched all "99" occurrences in the factory partition. Then I got:
0x4: xx:xx:xx:99:xx:80 (WiFi 2.4GHz)
0x2e: xx:xx:xx:99:xx:89 (WAN and LABEL)
0x8004: xx:xx:xx:99:xx:84 (WiFi 5GHz)
0xe000: xx:xx:xx:99:xx:88 (LAN)

I don't have w2914ns-v2 device, so it needs a test from someone else. But AFAIK the two devices are almost same, so I think it will work for w2914ns-v2 too.

adrianschmutzler replied:
Okay, can you check 0x28 and 0xe006 on your device, too?

mans0n replied:
It's blank.
0x28: fd:9b:ff:ff:ff:ff
0xe006: ff:ff:ff:ff:ff:ff

adrianschmutzler replied:
Okay, they really have a strange setup there, but what you did looks correct to me. Thank you for the additional research.

Would still be interested in the same information about w2914ns-v2 if someone can get it.

perillamint replied:
OK, I'll grab my factory of W2914NS-V2 today and will post the result of the analysis. today.

perillamint replied:
It is same as above result.

mans0n replied:
Great. That's the answer I've been waiting for. Thanks! :D

adrianschmutzler replied:
Okay, so this is settled. Since you have tested this, I would add this information to the commit message, like:
"
Both devices have an untypical MAC address configuration (both tested!):
0x4: xx:xx:xx:xx:xx:80 (WiFi 2.4GHz)
0x28: fd:9b:ff:ff:ff:ff (unused)
0x2e: xx:xx:xx:xx:xx:89 (WAN and label)
0x8004: xx:xx:xx:xx:xx:84 (WiFi 5GHz)
0xe000: xx:xx:xx:xx:xx:88 (LAN)
0xe006: ff:ff:ff:ff:ff:ff
"

mans0n replied:
Okay, added.

perillamint commented:
I did some experiments with WeVO W2914NS v2 firmware. I'll share some interesting things which I discovered.

First, WeVO uses the same header format which ipTIME(aka. TOTOLINK in other countries except S. Korea) products use. It checks the whole checksum of firmware file after 0x40 bytes, not only uImage.
Second, when I apply my mkiptimefw implementation on OpenWRT firmware, it refuses to boot.

mans0n commented:
@perillamint The stock firmware uses standard uImage format. The stock web UI refuses sysupgrade image, but will accept initramfs image. IMO we don't need special mkwevofw, but just a proper installation guide.

BTW could you please test this patch for w2914ns-v2 if you have one? https://github.com/openwrt/openwrt/pull/2336#discussion_r314983097

mans0n commented:
Rebased.

perillamint commented:
I can confirm stock bootloader of 11AC NAS Router happily accepts that uImage. However, It seems I lost W2914NS V2 U-Boot stock image. :(

@mans0n

mans0n commented:
@perillamint That's a shame... Thanks for your confirm on 11acnas, anyway. :)

mans0n commented:
Rebased to add label MAC.

mans0n commented:
I just realized that WiFi 2.4GHz LED is also broken (Why haven't I noticed that earlier?), so added another mt76 led node.

mans0n commented:
After a thorough review, I found a few more things to improve. Sorry for the frequent changes.
The commit is now quite big, do you think it should be split?

mans0n commented:
Fixed some typos.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_wevo_11acnas.dts:
```diff
@@ -6,10 +6,16 @@
+	aliases {
+		led-boot = &led_usb;
+		led-failsafe = &led_usb;
+		led-upgrade = &led_usb;
```
The LED used here should not be mistakable. Using USB, WiFi, LAN LEDs etc. is not a good idea.

So, either remove this entirely or at least make this a separate commit, so the rest is not delayed by this discussion.

mans0n replied:
Hmm... There's no device using USB LED for status indication so you may be right, but I still think it will be very helpful to have a status LED, at least during boot. I'll separate it as another commit then.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dts:
```diff
@@ -5,10 +5,16 @@
+	aliases {
+		led-boot = &led_usb;
+		led-failsafe = &led_usb;
+		led-upgrade = &led_usb;
```
Same as above.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_wevo_w2914ns-v2.dtsi:
```diff
@@ -11,19 +11,20 @@
 			label = "reset";
 			gpios = <&gpio0 29 GPIO_ACTIVE_LOW>;
 			linux,code = <KEY_RESTART>;
+			debounce-interval = <60>;
```
I thought debounce-interval is used for gpio-keys-polled?

mans0n replied:
gpio-keys also uses debounce-interval so it will be okay.
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/input/gpio-keys.txt#L25

adrianschmutzler commented:
> The commit is now quite big, do you think it should be split?

Yes, please. You do not need a separate commit for everything, but try to break it down into some logical units.

mans0n committed:
ramips: set uImage name of WeVO 11AC NAS and W2914NS v2

The stock firmware and bootloader only accept uImage with names that match
certain patterns. This patch enables OpenWrt installation from stock
firmware without having to reflash the bootloader or access the UART
console.

Installation via web interface:
1.  Flash **initramfs** image through the stock web interface.
2.  Boot into OpenWrt and perform sysupgrade with sysupgrade image.

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Split into three commits.

adrianschmutzler commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -427,6 +427,10 @@ wavlink,wl-wn575a3)
 	;;
+wevo,11acnas|\
+wevo,w2914ns-v2)
+	ucidef_set_led_usbport "usb" "USB" "$boardname:green:usb" "usb1-port1" "usb1-port2" "usb2-port1"
```
This seems redundant/overlapping with the triggers defined for the USB led. Do you need this to get back control after the diag.sh script was running?

mans0n replied:
Yes, exactly.

adrianschmutzler replied:
Okay, so this already shows why using the USB led for diag is not a good idea, as it requires "hacks" like this one. However, I'm not a core-developer, so I will leave this for other reviewers to decide.

mans0n replied:
In my defense, there's already some devices with status led that goes back to the original role after boot. https://github.com/openwrt/openwrt/commit/0c5a53255307ab532b2f875db452803b7f297874 https://github.com/openwrt/openwrt/commit/a9360452f0130bb39a8e732334a174cc3f8991fd https://github.com/openwrt/openwrt/commit/9f0e2335764674d0324dd90adb3e185cc8befe5f

adrianschmutzler replied:
Yes, and I would say it depends on the reviewer whether stuff like that comes in or not.

BTW, you provided an excellent choice of reference here, as each case is different:
0c5a532 uses LAN led without trigger in DTS
a936045 uses dts trigger for one led and 02_network for the other, but uses both for diag
9f0e233 uses signal strength indicator

Especially the last case has been accepted several times recently (particularly for the TP-Link CPExxx devices in ath79 target), as signal strenght indicators do not blink typically and thus have a reduced potential to confuse the user.
However, note that 9f0e233 uses the green indicator for diag AND signaling although the red one is completely unused. So this could also have slipped in ...

Nevertheless, I intentionally commented about this without a clear indication, as it depends on taste.

mans0n replied:
BTW "usbport" trigger does not blink, the LED just turns on when USB devices inserted. That's why I thought USB LED could also be a nice status LED.

adrianschmutzler replied:
No blinking on "activity"?

mans0n replied:
Nope. https://www.kernel.org/doc/html/latest/leds/ledtrig-usbport.html
"disk-activity" might blink, but "usbport" does not have any blink option AFAIK.

adrianschmutzler replied:
Okay, I'd say that this raises your chances :-)

If used as diag LED, one might consider removing the DTS trigger, so LED is only used for USB "after boot".

mans0n replied:
Good point. Done.

adrianschmutzler commented:
@mans0n Okay, I think the split is reasonably chosen.

@blocktrron Maybe have a look at this for review. In particular, we would need a verdict whether the approach in patch 3 is acceptable or not.

mans0n commented:
It seems like the `debounce-interval = <60>;` line has been originated from those old mach.c files in ar71xx and brcm63xx. In most mach files, the poll interval is set to 20ms and the debounce interval is set to three times the poll interval, which is 60ms.
I thought the longer debounce interval would help to reduce glitches, but after testing I found that the [default 5ms](https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/input/gpio-keys.txt#L26) debounce interval is enough for most use cases.
So I decided to remove `debounce-interval` from my PRs.

mans0n commented:
Rebased due to 19724e2.
What is blocking the merge of this PR?

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -199,6 +199,11 @@ ramips_setup_macs()
+	wevo,11acnas|\
+	wevo,w2914ns-v2)
+		wan_mac=$(mtd_get_mac_binary factory 0x2e)
+		label_mac=$wan_mac
```
I have just added a patch to master that sets lan+1 in an explicit case:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/board.d/02_network#L167

So, please rebase your patch onto that and _remove the wevo entries from this new node_.

adrianschmutzler replied:
I will merge this patch (patch 2) and rebase myself.

mans0n commented:
Rebased remaining commits.

adrianschmutzler commented:
Also update your initial comment here, maybe use strikethrough so the old stuff isn't just deleted.

mans0n commented:
Thanks for your advice.

mans0n commented:
I'm getting more and more negative about using USB LED for status indication (partly because I don't really use this device), so I'll withdraw the LED patch.
@adrianschmutzler Could you please merge the last remaining patch here? Actually that was the most essential part of this PR but it's been lingering too long.

adrianschmutzler commented:
Merged.
